qiskit/circuit/singleton_gate.py

```diff
@@ -0,0 +1,191 @@
+            )
+        self._label = label
+
+    @property
+    def condition(self):
+        return self._condition
+
+    @condition.setter
+    def condition(self, condition):
+        if self is self._instance:
+            raise NotImplementedError(
+                f"This gate class {type(self)} does not support manually setting a "
+                "condition on an instance. Instead you must set the label when instantiating a new "
+                "object or via the .c_if() method"
+            )
+        self._condition = condition
+
+    @property
+    def duration(self):
+        return self._duration
+
+    @duration.setter
+    def duration(self, duration):
+        if self is self._instance:
+            raise NotImplementedError(
+                f"This gate class {type(self)} does not support manually setting a "
+                "duration on an instance. Instead you must set the duration when instantiating a "
+                "new object."
+            )
+        self._duration = duration
+
+    @property
+    def unit(self):
+        return self._unit
+
+    @unit.setter
+    def unit(self, unit):
+        if self is self._instance:
+            raise NotImplementedError(
+                f"This gate class {type(self)} does not support manually setting a "
+                "unit on an instance. Instead you must set the unit when instantiating a "
+                "new object."
+            )
+        self._unit = unit
+
+    def __deepcopy__(self, _memo=None):
+        if (
+            self.condition is None
+            and self.label is None
+            and self.duration is None
+            and self.unit == "dt"
+        ):
+            return self
+        else:
+            return type(self)(
+                label=self.label, _condition=self.condition, duration=self.duration, unit=self.unit
+            )
+
+    def __setattr__(self, name, value):
+        if self.mutable:
+            super().__setattr__(name, value)
+        else:
+            if name not in {
+                "definition",
+                "unit",
+                "duration",
+                "condition",
+                "label",
+                "_label",
+                "_condition",
+                "_duration",
+                "_unit",
+                "_definition",
+                "_name",
+                "_num_qubits",
+                "_num_clbits",
+                "_params",
+                "params",
+            }:
+                raise NotImplementedError(
+                    "Setting custom attributes is not allowed on a singleton gate"
+                )
+            super().__setattr__(name, value)
+
+    def copy(self, name=None):
+        if name is not None and self.condition is None and self.label is None:
+            raise QiskitError("A custom name can not be set on a copy of a singleton gate")
+        return super().copy()
```

Comment:
This doesn't quite feel right - it feels like it should be more like
```python
def copy(self, name=None):
    if not self.mutable and name is None:
        return self
    return super().copy(name=name)  # Or whatever special handling for `name`.
```

At the moment it feels like there's a logic problem if `name is not None` but we're in mutable mode - it doesn't get propagated.

Reply:
TBH, I don't remember the rational for the logic here and it does look weird (it might have just been leftover from an earlier iteration and I missed updating it). Your suggestion looks correct so I'll switch it to that.

Reply:
Fixed in: https://github.com/Qiskit/qiskit/pull/10314/commits/4b242e12eb8da7dd13b851cef21b79cd111a051d